OPTM-67 Improve QR code display

guisedlythe opened the pull request:
We might also want to remove the IP selection from the Settings screen.

OPTM-67 #close

guisedlythe committed:
OPTM-67 Improved QR display screen

- Added IP selection
- Added port display
- Added warning if server is not running

guisedlythe committed:
OPTM-76 Show dialog asking to start server if it is not running when opening QR screen.

OPTM-76 #close

guisedlythe committed:
OPTM-76 Don't show QR screen button when using web

guisedlythe committed:
Merge remote-tracking branch 'origin/main' into OPTM-67-qr-code

firstdayofjune commented on lib/screens/home_screen.dart:
```diff
@@ -58,4 +59,52 @@ class _HomeScreenState extends State<HomeScreen> {
+        Provider.of<ChangeCallbackNotifier<PeerService>>(context, listen: false)
+            .callbackProvider;
+    if (!peerService.isServerRunning) {
+      showDialog(
```
This is a relatively long method, which moves the else statement "out of sight" from the if.
Please move the showDialog into a separate method and call this method in the if block instead.
This increases:

-  readability (because the if condition is only separated from the else by a single line)
- testability (because the method building the dialog con be tested on its own

firstdayofjune commented on lib/screens/home_screen.dart:
```diff
@@ -58,4 +59,52 @@ class _HomeScreenState extends State<HomeScreen> {
+                  builder: (context) => QrCodeDialog(),
+                );
+              },
+              child: Text('No'),
```
Please change the code to close the modal, if a user selects "No".

Rational: I was confused to see the QR Code display even after selecting "No". Also, I don't see the value in displaying the QR Code without an active server.

firstdayofjune commented on lib/screens/qr_code_dialog.dart:
```diff
@@ -35,19 +46,68 @@ class QrCodeDialog extends StatelessWidget {
+            }
+
+            return Column(children: [
+              if (!peerService.isServerRunning)
```
If the QR Code cannot be opened without a running server (see comment above), this check may be removed.

firstdayofjune commented on lib/screens/qr_code_dialog.dart:
```diff
@@ -35,19 +46,68 @@ class QrCodeDialog extends StatelessWidget {
+                ),
+              Row(mainAxisAlignment: MainAxisAlignment.center, children: [
+                Text('IP: '),
+                DropdownButton<String>(
```
If there is only a single IP Address available, display the IP Address as text (like the port) instead of as a dropdown.

firstdayofjune reviewed (changes requested):
Looks good. Only some minor changes, as described in the comments.

guisedlythe committed:
OPTM-67 Integrated most PR review comments.

- Some refactoring. Split QrCodeDialog.build to multiple methods to reduce method length.
- Don't display drop down if there is only one IP available, use simple text instead.
- Show warning when no IP could be found.
- Changed alert dialog options when server is not running to "Yes", "Cancel", "Show anyway"

guisedlythe commented:
**Regarding the alert dialog options:**
I changed the dialog options to "Yes", "Cancel", and "Show anyway" (c.f. bd7157d)
Reasoning:
I wanted to keep the option to show the QR code even if the server is not running. While you cannot connect to the device if the server is not running, you can still add it to the list of known devices. When you then start the server later, it will be used in sync. I understand that this is an edge case, but I generally dislike removing features, making it overly difficult to do what you want in some cases. I didn't add a third option before, because you can just click anywhere else on the screen to close the dialog.
